artemis-commons/src/main/java/org/apache/activemq/artemis/core/server/ActiveMQScheduledComponent.java

```diff
@@ -287,35 +295,47 @@ private void restartIfNeeded() {
       }
    }
 
-   final Runnable runForExecutor = new Runnable() {
-      @Override
-      public void run() {
-         if (onDemand && delayed.get() > 0) {
-            delayed.decrementAndGet();
-         }
+   private void runForExecutor(AtomicBoolean booked) {
+      // It unblocks:
+      // - a new delay request
+      // - next periodic run request (in case of executor != null)
+      // Although tempting, don't move this one after ActiveMQScheduledComponent.this.run():
+      // - it can cause "delay" to change semantic ie a racing delay while finished executing the task, won't succeed
+      // - it won't prevent "slow tasks" to accumulate, because slowness cannot be measured inside running method;
+      //   it just cause skipping runs for perfectly timed executions too
+      // Re the code as it is: if the next periodic run has been able to request a new round while
+      // the previous one hasn't yet started, it makes sense to skip it, because its job will be executed by
+      // the old one (yet to start).
```

Comment:
Not sure this bit of comment is in the right place, or if so could do with updating. This code here doesnt skip anything, as the comment reads like. It instead resets the flag used by other areas to skip.

Reply:
Yep effectively is not the right place for this bits of code, but is more the overall logic...I can put it in the header of the class docs!